src/hooks/useEvidences.js

```diff
@@ -30,11 +27,19 @@ export default function useEvidences(evidences) {
 
       // If is a valid CID try to fetch the content
       if (ipfsCid) {
+        // If the cid was already processed exit
+        if (evidenceCache.get(ipfsCid)) {
```

Comment:
So asynchronous operations are always a bit tricky, but moving the logic to be here is less safe than before.

What can happen now:

1. In one render, start fetching `cidA`
2. Quickly afterwards, we re-render, and we again start fetching `cidA`
  - Note here that if we're still waiting to hear back from `await ipfsGet('cidA')`, `evidenceCache` won't have been set yet

This can't happen right now, but is really easy to become a possibility if any action after `evidenceCache.set()` becomes async:

1. In one render, start fetching `cidA`
2. Continue after `await ipfsGet('cidA')`
3. Cancel the render, because we re-render (this requires an async action to happen after `evidenceCache.set()`)
4. Now on the next renders, because we've already set the evidence cache, we skip this value and it will never appear.

Reply:
For the sake of sanity and being safe, let's:

- Remove the caching idea; it adds too much complexity to do correctly, and is not valuable enough here
- Make sure `id` is in each object of `evidenceProcessed` (I believe we need to add this in just the error cases)
- Before we do `evidences.forEach()`, let's filter the `evidences` list by what we already have in `evidenceProcessed`, and update `evidenceProcessed` to filter out any items that aren't in `evidences`
  - In set terminology:  `Set(evidences) - Set(evidenceProcessed)`, and `Set(evidences) U Set(evidenceProcessed)`
  - Then do the `forEach()` over only the new evidences (`Set(evidences) - Set(evidenceProcessed)`)

In this scenario, if we've protected ourselves correctly with `cancelled` guarding any state setters (i.e. `setEvidenceProcessed()`), we should never have a case where a re-render would cause inconsistent state. We only "commit" state when we're in the same render, and we only fetch new evidences when they're not in the "committed" state

Reply:
What about:

- Having a single evidence fetching function, that only caches valid evidences (non errored).
- Move the fetching + transformation + caching + error state in this function.
- Only fetch all the evidences + update the result in the state in `useEffect()`.
- Update all the evidences at the same time rather than appending.

/!\ Untested code /!\

```jsx
function useEvidences(rawEvidences) {
  // Contains valid evidences + errored evidences
  const [evidences, setEvidences] = useState([])

  // Contains valid evidences only
  const evidencesCache = useRef(new Map())

  // Fetch an evidence data from IPFS if needed, prepares the evidence object,
  // and cache it if valid. If invalid, returns an errored evidence object.
  const fetchEvidence = useCallback(async rawEvidence => {
    const { id, data: uriOrData, submitter, createdAt } = rawEvidence

    if (ipfsCache.current.has(id)) {
      return ipfsCache.current.get(id)
    }

    const baseEvidence = {
      id,
      metadata: null,
      defendant: '',
      agreementText: '',
      submitter,
      createdAt,
      error: false,
    }

    // We might want to rename this function `getIpfsCidFromUri()` or `parseIpfsUri()`
    const cid = getIpfsCidFromString(uriOrData)

    // Not an IPFS URI
    if (!cid) {
      ipfsCache.current.set(id, { ...baseEvidence, metadata: uriOrData })
      return ipfsCache.current.get(id)
    }

    const { data, error } = await ipfsGet(cid)

    if (error) {
      return { ...baseEvidence, error: ERROR_TYPES.ERROR_FETCHING_IPFS }
    }

    // we might re-set it here, but it’s not really a problem
    ipfsCache.current.set(id, data)
    return {
      ...baseEvidence,
      metadata: data,
    }
  }, [])

  useEffect(() => {
    let cancelled = false

    const updateEvidences = async () => {
      const evidences = await Promise.all(rawEvidences.map(fetchEvidence))
      if (!cancelled) {
        setEvidences(evidences.sort((a, b) => a.createdAt - b.createdAt))
      }
    }

    updateEvidences()

    return () => {
      cancelled = true
    }
  }, [rawEvidences, fetchEvidence])
}
```

Reply:
With that promise all, how are we going to handle the timeout for the request? it can't be too much time right? because we can be loading disputes for a long time with the user waiting to load all the disputes

Reply:
> With that promise all, how are we going to handle the timeout for the request?

This should be handled by `ipfsGet()`; if it times out, we return an error.